jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java

```diff
@@ -980,7 +978,10 @@ public void demand(Runnable demandCallback)
                 {
                     if (httpChannelState._onContentAvailable != null)
                         throw new IllegalArgumentException("demand pending");
-                    httpChannelState._onContentAvailable = demandCallback;
+                    if (demandCallback instanceof Invocable)
+                        httpChannelState._onContentAvailable = demandCallback;
+                    else
+                        httpChannelState._onContentAvailable = new ReadyTask(InvocationType.NON_BLOCKING, demandCallback);
```

Comment:
This is the core of this alternative proposal....  but I don't think it has achieved much as there are still 22 other files that are modified as you attempt to pass an Invocable rather than a plain Runnable.  So this PR is making pretty much the same changes as the original anyway.  

Changing the contract here doesn't win very much, nor does it help with the other 43 implementations of demand that all need to be checked and/or modified.

Reply:
It is not about the change we make to Jetty files.

It is about the changes that applications need to do.
For example, with the other PR, CometD would need to be changed. With this PR, no.

Reply:
Can you point to the changes needed in CometD?   Are they really needed?  Are there demand callbacks in cometd that are tagged as blocking, but ultimately do nothing but wake up another blocked thread?   If not, then they are not a problem.

The invocation type of our callbacks has always key, but it is only vital if it is used in the conversion from async to blocking APIs, which is typically done by our utility classes.    The fact that we messed up a few utility classes and forgot to pass on the invocation type should not mean that we change the contract.  If we assume non-blocking and call blocking code, we will starve the thread pool just the same.

Reply:
@sbordet the more I think about this, the more I'm convinced this is not an issue with the `demand` API/contract.    There is nothing wrong with passing blocking tasks to `demand` and there is nothing wrong with assuming blocking unless told otherwise.  The mistake we made in a few recent utility classes was to block a Thread and to wait for a wakeup from a blocking task that was likely to come from the same thread pool as the original blocked thread.    This is just a basic programming error and the fact that the `demand` API was used is not really relevant, as the same issue would occur if `Executor.execute` was used.

Our previous/existing utility classes (`Blocker`, `SharedBlockingCallback`, `BlockingCallback`, `HttpInput` etc.) were all written correctly with the knowledge that if you block a thread, then you need to be very careful about how the event that wakes it up is scheduled and that it cannot be dependent on the wakeup of the blocked thread.

The fact that we made some programming errors in our refactor of `HttpInput` and the new `CompletableFuture` based utilities is on us, and not on the `demand` API.   We made some error, which are easy to fix (done in the original PR).     Sure others can make similar errors: async programming is hard.  But changing the contract of `demand` so our errors are glossed over, is not a fix.  It actually makes it worse, as we may now dispatch the last available thread to a callback that is validly blocking, but that we decided to declare non-blocking without any indication to the caller.... just on the assumption that if they happened to make the same programming mistake that we did, then we'd work around it.  If they have not made the same mistake, then this PR will lock the server.